pgx/src/datum/time_stamp_with_timezone.rs

```diff
@@ -36,45 +37,54 @@ impl FromDatum for TimestampWithTimeZone {
         if is_null {
             None
         } else {
-            let mut tm = pg_sys::pg_tm {
-                tm_sec: 0,
-                tm_min: 0,
-                tm_hour: 0,
-                tm_mday: 0,
-                tm_mon: 0,
-                tm_year: 0,
-                tm_wday: 0,
-                tm_yday: 0,
-                tm_isdst: 0,
-                tm_gmtoff: 0,
-                tm_zone: std::ptr::null_mut(),
-            };
-            let mut tz = 0i32;
-            let mut fsec = 0 as pg_sys::fsec_t;
-            let mut tzn = std::ptr::null::<std::os::raw::c_char>();
-            pg_sys::timestamp2tm(
-                datum.value() as i64,
-                &mut tz,
-                &mut tm,
-                &mut fsec,
-                &mut tzn,
-                std::ptr::null_mut(),
-            );
-            let date = time::Date::from_calendar_date(
-                tm.tm_year,
-                time::Month::try_from(tm.tm_mon as u8)
-                    .expect("Got month outside of range in TimestampWithTimeZone::from_datum"),
-                tm.tm_mday as u8,
-            )
-            .expect("failed to create date from TimestampWithTimeZone");
+            let datum_val = datum.value() as i64;
+            let (date, time, tz) = match datum_val {
+                i64::MIN => (time::Date::MIN, time::Time::MIDNIGHT, 0i32),
```

Comment:
The only questionable decision i made was just setting the date portion to MIN/MAX and not the time. is it worth making the infinity timestamp have a time of 23:59:59.999?

Reply:
Yes. If we use the MIN/MAX values that's one thing but it should be an unambiguous comparison for the multi-component case, so that when negative infinity is mapped, all other values are greater-or-equal, and when positive infinity is mapped, all other values are lesser-or-equal.

Reply:
I am now reminded that postgres can represent dates way higher than `time::Date::MAX` `9999-12-31 23:59:59.999999` ex. pg dates go from 4713 BC to 5874897 AD and timestamp from 4713BC to 294276 AD.  So we could easily break the current mapping with just `date'10001-01-01'` I've heard `chrono` has a wider range, but is it possible to map to a rust enum?

```
enum PgDate {
    NegInfinity,
    OutOfRange(i64), // epoch days do what you will
    Date(time::Date),
    Infinity
}
```